src/Course/State.hs

```diff
@@ -135,44 +122,137 @@ instance Monad (State s) where
 -- prop> \xs -> case firstRepeat xs of Empty -> True; Full x -> let (l, (rx :. rs)) = span (/= x) xs in let (l2, r2) = span (/= x) rs in let l3 = hlist (l ++ (rx :. Nil) ++ l2) in nub l3 == l3
-firstRepeat ::
-  Ord a =>
-  List a
-  -> Optional a
-firstRepeat =
-  error "todo: Course.State#firstRepeat"
+--
+-- attempt 1: my misconception of what was being asked because I couldn't
+-- read the test cases
+--
+-- firstRepeat :: Ord a => List a -> Optional a
+-- firstRepeat Nil = Empty
+-- firstRepeat (_ :. Nil) = Empty
+-- firstRepeat (x1 :. (x2 :. xs))
+--   | x1 == x2 = Full x1
+--   | otherwise = firstRepeat (x2 :. xs)
+
+-- attempt 2: just doin' it my way, worked first time
+--
+-- firstRepeat :: Ord a => List a -> Optional a
+-- firstRepeat xs = firstRepeat1 xs S.empty
+--   where
+--     firstRepeat1 :: Ord a => List a -> S.Set a -> Optional a
+--     firstRepeat1 Nil _ = Empty
+--     firstRepeat1 (x :. rest) set
+--       | S.member x set = Full x
+--       | otherwise = firstRepeat1 rest (S.insert x set)
+
+-- attempt 3: using findM and a State
+--
+firstRepeat :: Ord a => List a -> Optional a
+firstRepeat xs = eval (findM p xs) S.empty
+  where
+    p x = State $ \set -> (S.member x set, S.insert x set)
+
+----------------------------------------------------
+----------------------------------------------------
+----------------------------------------------------
+-- debugging with ints
 
+traceMonad :: (Show a, Monad m) => [Char] -> a -> m a
+traceMonad str x = trace (str P.++ show x) (return x)
+
+findMN :: Monad f => (Int -> f Bool) -> List Int -> f (Optional Int)
+findMN p xs = foldRight (findMNFold p) (return Empty) xs
+
+findMNFold :: Monad f => (Int -> f Bool) -> Int -> f (Optional Int) -> f (Optional Int)
+findMNFold p x result =
+      p x
+      >>= \b -> if b then return (Full x) else result
+      >>= traceMonad "result = "
+
+firstRepeatN :: List Int -> Optional Int
+firstRepeatN xs = fst $ runState (findMN p xs) S.empty
+  where
+    p x | trace ("firstRepeat p x, x = " P.++ show x) False = undefined
+    p x = get >>= \set -> State $ \xyz -> (S.member x (trace (show xyz) set), S.insert x xyz)
+
+memberAndAddN :: Int -> State (S.Set Int) Bool
+memberAndAddN x | trace ("memberAndAddN x = " P.++ show x) False = undefined
+memberAndAddN x = State (\set -> memberAndAddSetN x set)
+
+memberAndAddSetN :: Int -> S.Set Int -> (Bool, S.Set Int)
+memberAndAddSetN x set | trace ("memberAndAddSetN x = " P.++ show x P.++ ", set = " P.++ show set) False = undefined
+memberAndAddSetN x set = (S.member x set, S.insert x set)
+
+--
+-- at ghci
+-- let p x = get >>= \s -> put (Data.Set.insert x s) >>= \_ -> return (Data.Set.member x s) in runState (findM p (listh [1,2,3,1,2,3])) Data.Set.empty
+--
+-- simpler without bind:
+--
+-- let p x = State (\set -> (Data.Set.member x set, Data.Set.insert x set)) in runState (findM p (listh [1,2,3,1,2,3])) Data.Set.empty
```

Comment:
this is really elegant :) I think previous greg might have found the monadic version easier to read (and maybe current Julian does, too)... but it's good either way.